File: command/mycommand.cpp
#include <iostream>

using namespace std;

class Action {
    public:
    virtual ~Action(){}
    virtual void DoIt() const = 0;
};

class Peel : public Action {
    private:
    string vegetable;

    public:
    explicit Peel(string vegetable) : vegetable(vegetable) {
    }

    void DoIt() const override {
        cout << "Peeling the vegetable " << this->vegetable << endl;
    }
};

class Customer {
    public:
    void OrderFood(const string &a) {
    }
    void Pay(const string &a) {
    }
};

class InteractWithCustomer: public Action {
    private:
    Customer *customer;

    string a;

    public:
    InteractWithCustomer(Customer *customer, string a) : 
        customer(customer), a(a) {
        }
    void DoIt() const override {
        this->customer->OrderFood(this->a);
        this->customer->Pay(this->a);
    }
};

class Waiter {
    private:
    Action *action1;

    Action *action2;

    public:
    ~Waiter() {
        delete action1;
        delete action2;
    }

    void DoAction1(Action *action) {
        this->action1 = action;
    }
    void DoAction2(Action *action) {
        this->action2 = action;
    }

    void Perform() {
        if (this->action1) {
            this->action1->DoIt();
        }
        if (this->action2) {
            this->action2->DoIt();
        }
    }
};

int main() {
    Waiter *waiter = new Waiter();
    waiter->DoAction1(new Peel("Potato"));

    Customer *customer = new Customer();
    waiter->DoAction2(new InteractWithCustomer(customer, "How Can I Help You!"));
    waiter->Perform();

    delete waiter;
    delete customer;

    return (0);
}

File: command/swift/mycommand.swift
import Foundation

// 1. Action (Command) Interface
protocol Action {
    func doit()
}

// 2. Receiver Class
class Customer {
    let name: String
    
    init(name: String) {
        self.name = name
    }
    
    func orderFood() {
        print("\(name) is ordering food.")
    }
    
    func makePayment() {
        print("\(name) is making a payment.")
    }
}

// 3. Concrete Command for Customer Interaction
class CustomerInteraction: Action {
    private let customer: Customer
    private let interactionType: String
    
    init(customer: Customer, type: String) {
        self.customer = customer
        self.interactionType = type
    }
    
    func doit() {
        if interactionType == "order" {
            customer.orderFood()
        } else if interactionType == "pay" {
            customer.makePayment()
        }
    }
}

// 4. Concrete Command for Peeling
class PeelAction: Action {
    let item: String
    
    init(item: String) {
        self.item = item
    }
    
    func doit() {
        print("Peeling the \(item).")
    }
}

// 5. Invoker Class
class Waiter {
    private var actions: [Action] = []
    
    func addAction(_ action: Action) {
        actions.append(action)
    }
    
    func executeActions() {
        print("Waiter is processing the queue...")
        actions.forEach { $0.doit() }
        actions.removeAll()
    }
}

// 6. Main Implementation
let client = Customer(name: "John Doe")

// Initialize specific actions
let orderAction = CustomerInteraction(customer: client, type: "order")
let payAction = CustomerInteraction(customer: client, type: "pay")
let peelingPotato = PeelAction(item: "Potato")

// Set up the Waiter (Invoker)
let waiter = Waiter()
waiter.addAction(orderAction)
waiter.addAction(peelingPotato)
waiter.addAction(payAction)

// Execute all commands
waiter.executeActions()
